File: Core/Catalogues/CollectionCatalogue.cs
using KolibSoft.Catalogue.Core.Abstractions;
using KolibSoft.Catalogue.Core.Utils;
using static KolibSoft.Catalogue.Core.Constants;

namespace KolibSoft.Catalogue.Core.Catalogues;

public class CollectionCatalogue<TItem, TFilters> : ICatalogueConnector<TItem, TFilters>
    where TItem : class, IItem, IValidatable, IUpdatable<TItem>
{

    public ICollection<TItem> Collection { get; } = new List<TItem>();
    public ICollection<string> Errors { get; } = new List<string>();
    public virtual bool Available => true;

    protected virtual Task<IEnumerable<TItem>> OnQueryAsync(TFilters? filters) => Task.FromResult<IEnumerable<TItem>>(Collection);
    protected virtual Task<TItem?> OnGetAsync(Guid id) => Task.Run(() => Collection.FirstOrDefault(x => x.Id == id));
    protected virtual Task<bool> OnValidateInsertAsync(TItem item) => Task.FromResult(true);
    protected virtual Task<bool> OnValidateUpdateAsync(TItem item) => Task.FromResult(true);
    protected virtual Task<bool> OnValidateDeleteAsync(TItem item) => Task.FromResult(true);

    public virtual async Task<Result<Page<TItem>?>> QueryAsync(TFilters? filters = default, int pageIndex = 0, int pageSize = DefaultChunkSize)
    {
        Errors.Clear();
        var items = await OnQueryAsync(filters);
        var page = items.GetPage(pageIndex, pageSize);
        return page;
    }

    public virtual async Task<Result<TItem?>> GetAsync(Guid id)
    {
        Errors.Clear();
        var item = await OnGetAsync(id);
        return item;
    }

    public virtual async Task<Result<TItem?>> InsertAsync(TItem item)
    {
        Errors.Clear();
        if (!item.Validate(Errors)) return Errors.ToArray();
        var original = Collection.FirstOrDefault(x => x.Id == item.Id);
        if (original != null) return default(TItem?);
        if (!await OnValidateInsertAsync(item)) return Errors.ToArray();
        Collection.Add(item);
        return item;
    }

    public virtual async Task<Result<TItem?>> UpdateAsync(Guid id, TItem item)
    {
        Errors.Clear();
        if (!item.Validate(Errors)) return Errors.ToArray();
        var original = Collection.FirstOrDefault(x => x.Id == id);
        if (original == null) return default(TItem?);
        if (!await OnValidateUpdateAsync(item)) return Errors.ToArray();
        if (item.ModifiedAt >= original.ModifiedAt) original.Update(item);
        return original;
    }

    public virtual async Task<Result<TItem?>> DeleteAsync(Guid id)
    {
        Errors.Clear();
        var item = Collection.FirstOrDefault(x => x.Id == id);
        if (item == null) return default(TItem?);
        if (!await OnValidateDeleteAsync(item)) return Errors.ToArray();
        Collection.Remove(item);
        return item;
    }

    public CollectionCatalogue(ICollection<TItem>? collection = default)
    {
        Collection = collection ?? new List<TItem>();
    }

}

File: docs/lib/catalogues/collection_catalogue.js
import { CatalogueConnector } from "../abstractions/catalogue_connector.js";
import { DefaultChunkSize } from "../constants.js";
import { Page } from "../page.js";
import { Result } from "../result.js";
import { PageUtils } from "../utils/page_utils.js";

/**
 * @template TItem 
 * @template TFilters
 * @implements {CatalogueConnector<TItem, TFilters>}
 */
class CollectionCatalogue extends CatalogueConnector {

    /** @type {TItem[]} */
    #collection;

    /** @type {string[]} */
    #errors;

    /** @returns {TItem[]} */
    get collection() { return this.#collection; }

    /** @returns {string[]} */
    get errors() { return this.#errors; }

    /** @returns {boolean} */
    get available() { return true; }

    /**
     * @param {TFilters?} filters 
     * @returns {TItem[]}
     */
    async onQueryAsync(filters) { return this.#collection; }

    /**
     * @param {string} id 
     * @returns {TItem?}
     */
    async onGetAsync(id) { return this.#collection.find(x => x.id == id); }

    /**
     * @param {TItem} item 
     * @returns {boolean}
     */
    async onValidateInsertAsync(item) { return true; }

    /**
     * @param {TItem} item 
     * @returns {boolean}
     */
    async onValidateUpdateAsync(item) { return true; }

    /**
     * @param {TItem} item 
     * @returns {boolean}
     */
    async onValidateDeleteAsync(item) { return true; }

    /**
     * @param {TFilters?} filters 
     * @param {number} pageIndex 
     * @param {number} pageSize 
     * @returns {Promise<Result<Page<TItem>?>>}
     */
    async queryAsync(filters = null, pageIndex = 0, pageSize = DefaultChunkSize) {
        this.#errors = [];
        let items = await this.onQueryAsync(filters);
        let page = PageUtils.getPage(items, pageIndex, pageSize);
        return new Result({ data: page });
    }

    /**
     * @param {string} id 
     * @returns {Promise<Result<TItem?>>}
     */
    async getAsync(id) {
        this.#errors = [];
        let item = await this.onGetAsync(id);
        return new Result({ data: item });
    }

    /**
     * @param {TItem} item 
     * @returns {Promise<Result<TItem?>>}
     */
    async insertAsync(item) {
        this.#errors = [];
        if (!item.validate(this.#errors)) return new Result({ errors: this.#errors });
        let original = this.#collection.find(x => x.id == item.id);
        if (original != null) return new Result({ data: null });
        if (!(await this.onValidateInsertAsync(item))) return new Result({ errors: this.#errors });
        this.#collection.push(item);
        return new Result({ data: item });
    }

    /**
     * @param {string} id 
     * @param {TItem} item 
     * @returns {Promise<Result<TItem?>>}
     */
    async updateAsync(id, item) {
        this.#errors = [];
        if (!item.validate(this.#errors)) return new Result({ errors: this.#errors });
        let original = this.#collection.find(x => x.id == id);
        if (original == null) return new Result({ data: null });
        if (!(await this.onValidateUpdateAsync(item))) return new Result({ errors: this.#errors });
        original.update(item);
        return new Result({ data: original });
    }

    /**
     * @param {string} id 
     * @returns {Promise<Result<TItem?>>}
     */
    async deleteAsync(id) {
        this.#errors = [];
        let item = this.#collection.find(x => x.id == item.id);
        if (item == null) return new Result({ data: null });
        if (!(await this.onValidateDeleteAsync(item))) return new Result({ errors: this.#errors });
        this.#collection.splice(this.#collection.indexOf(item), 1);
        return new Result({ data: item });
    }

    /**
     * @param {TItem[]?} collection 
     */
    constructor(collection = null) {
        super();
        this.#collection = collection ?? [];
        this.#errors = [];
    }

}

export {
    CollectionCatalogue
}